gapis/resolve/dependencygraph2/graph_visualization/graph_structure.go

```diff
@@ -257,60 +265,111 @@ func (g *graph) makeStronglyConnectedComponentsByCommandTypeId() {
 	}
 }
 
-func (g *graph) getEdgesInDotFormat() string {
-	output := ""
-	for _, currentEdge := range g.edgeIdToEdge {
-		lines := strconv.Itoa(currentEdge.source.id) + " -> " + strconv.Itoa(currentEdge.sink.id) + ";\n"
-		output += lines
+func (g *graph) dfs(currentNode *node, visited []bool, visitedNodes *[]*node) {
+	*visitedNodes = append(*visitedNodes, currentNode)
+	visited[currentNode.id] = true
+	neighbours := g.getSortedNeighbours(currentNode.outNeighbourIdToEdgeId)
+	for _, neighbour := range neighbours {
+		if !visited[neighbour.id] {
+			g.dfs(neighbour, visited, visitedNodes)
+		}
+	}
+
+	for _, subCommandNode := range currentNode.subCommandNodes {
+		if !visited[subCommandNode.id] {
+			g.dfs(subCommandNode, visited, visitedNodes)
+		}
 	}
-	return output
 }
 
-func (g *graph) getNodesInDotFormat() string {
-	output := ""
-	for _, currentNode := range g.nodeIdToNode {
-		lines := strconv.Itoa(currentNode.id) + "[label=" + currentNode.label + "]" + ";\n"
-		output += lines
+func (g *graph) joinNodesByFrame() {
+	visited := make([]bool, g.maxNodeId+1)
+	frameNumber := 1
+	nodes := g.getSortedNodes()
+	for _, currentNode := range nodes {
+		if !visited[currentNode.id] && currentNode.isEndOfFrame {
```

Comment:
Just an observation: Forward dependencies break this strategy. For example, if an image is created in frame 1 and destroyed in frame 2, the creation has a dependency to the destruction. Following that edge in the DFS would cause some of the frame 2 commands to be incorrectly placed in frame 1.

However, I think the issues from this should be pretty small, since nodes with forward dependencies (like resource destruction) probably don't have a lot of other dependencies. I also don't see an obviously more correct way to handle forward dependencies, since just ignoring them in the DFS will leave lots of orphan nodes not associated with any frame.